instrumentation/aws-lambda-1.0/library/src/main/java/io/opentelemetry/instrumentation/awslambda/v1_0/TracingRequestApiGatewayWrapper.java

```diff
@@ -5,23 +5,58 @@
 
 package io.opentelemetry.instrumentation.awslambda.v1_0;
 
+import com.amazonaws.services.lambda.runtime.Context;
 import com.amazonaws.services.lambda.runtime.events.APIGatewayProxyRequestEvent;
 import com.amazonaws.services.lambda.runtime.events.APIGatewayProxyResponseEvent;
+import com.fasterxml.jackson.core.JsonProcessingException;
 import io.opentelemetry.sdk.OpenTelemetrySdk;
+import java.util.function.BiFunction;
 
 /**
  * Wrapper for {@link TracingRequestHandler}. Allows for wrapping a lambda proxied through API
  * Gateway, enabling single span tracing and HTTP context propagation.
  */
 public class TracingRequestApiGatewayWrapper
-    extends TracingRequestWrapperBase<APIGatewayProxyRequestEvent, APIGatewayProxyResponseEvent> {
+    extends TracingRequestWrapperBase<APIGatewayProxyRequestEvent, Object> {
 
   public TracingRequestApiGatewayWrapper() {
-    super();
+    super(TracingRequestApiGatewayWrapper::map);
   }
 
   // Visible for testing
-  TracingRequestApiGatewayWrapper(OpenTelemetrySdk openTelemetrySdk, WrappedLambda wrappedLambda) {
-    super(openTelemetrySdk, wrappedLambda);
+  TracingRequestApiGatewayWrapper(
+      OpenTelemetrySdk openTelemetrySdk,
+      WrappedLambda wrappedLambda,
+      BiFunction<APIGatewayProxyRequestEvent, Class, Object> mapper) {
+    super(openTelemetrySdk, wrappedLambda, mapper);
+  }
+
+  // Visible for testing
+  static Object map(APIGatewayProxyRequestEvent event, Class clazz) {
+    try {
+      return OBJECT_MAPPER.readValue(event.getBody(), clazz);
+    } catch (JsonProcessingException e) {
+      throw new IllegalStateException(
+          "Could not map API Gateway event body to requested parameter type: " + clazz, e);
+    }
+  }
+
+  @Override
+  protected APIGatewayProxyResponseEvent doHandleRequest(
+      APIGatewayProxyRequestEvent input, Context context) {
+    Object result = super.doHandleRequest(input, context);
+    APIGatewayProxyResponseEvent event = null;
+    // map to response event if needed
+    if (result instanceof APIGatewayProxyResponseEvent) {
+      event = (APIGatewayProxyResponseEvent) result;
+    } else {
+      try {
+        event = new APIGatewayProxyResponseEvent();
+        event.setBody(OBJECT_MAPPER.writeValueAsString(result));
+      } catch (JsonProcessingException e) {
+        throw new IllegalStateException("Could not serialize return value.", e);
```

Comment:
These exceptions seem worrisome, are we sure we won't break the app?

Reply:
The silver lining here is that if the wrapper is unable to map to a custom type, the user can still use the "APIGatewayProxyRequestEvent" and avoid mapping on the wrapper side entirely. So in the worst case it will work exactly the same way it does right now.

Reply:
I think it still means we have the possibility of breaking the lambda just by the presence of a lambda layer such as the `opentelemetry-lambda` ones. I don't think it's good to have the possibility of adding crashing.

That being said I guess I can't find any alternative unfortunately. It's too bad the Context object doesn't provide information about HTTP requests...

Is it at least beneficial to split into one where the target handler is proxy that doesn't have any mapping and this one that does for custom return type?

Reply:
It's implemented (kind of) this way - mapping ins only done if the return type of the handler is different than APIGatewayProxyResponseEvent. So if the wrapper mapping causes issues, the user can always switch to return the event. In other words both with input and result if the new wrapper breaks, user can implement own handler to input / return the event - and everything will work the same way it did (no mapping if no custom type used).